test/external-modules/apm-integration/src/javaRestTest/java/org/elasticsearch/test/apmintegration/RecordingApmServer.java

```diff
@@ -75,21 +74,27 @@ private Thread consumerThread() {
     @Override
     protected void after() {
         server.stop(1);
-        consumerRunning = false;
+        consumer = null;
+    }
+
+    void stop() {
+        running = false;
     }
 
     private void handle(HttpExchange exchange) throws IOException {
         try (exchange) {
-            try {
-                try (InputStream requestBody = exchange.getRequestBody()) {
-                    if (requestBody != null) {
-                        var read = readJsonMessages(requestBody);
-                        received.addAll(read);
+            if (running) {
+                try {
+                    try (InputStream requestBody = exchange.getRequestBody()) {
+                        if (requestBody != null) {
+                            var read = readJsonMessages(requestBody);
+                            received.addAll(read);
+                        }
                     }
-                }
 
-            } catch (RuntimeException e) {
-                logger.warn("failed to parse request", e);
+                } catch (RuntimeException e) {
```

Comment:
Wondering, we could also handle the IOException here?
Currently we just swallow runtime exceptions, we could capture any exception and assert there hasn't been any at the end of the test.... and whatever happens afterwards doesn't matter.
To me that feels a bit more robust than having to know that the mock server resource has to be stopped in advance - and understanding why that doesn't hurt ES

Reply:
I was wondering the same; we can just handle Throwable and don't care (log, but don't care):  if we have an error in communicating to/from the mock server the test would fail anyway as the expected messages will not arrive.
I preferred the "stop" thing as it solves the issue, but I'm not sure.